[Feat] 상품 주문

yoonnable opened the pull request:
## 🌿 PR타입
- [x] 기능 추가
- [x] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
- [x] 기타 사소한 수정

## ❗️관련 이슈[close #26 ]
### 📑 개요
* 상품 주문 기능을 구현합니다.
* 상품 상세 페이지에서의 즉시 구매와 장바구니에서 선택한 상품 구매가 가능합니다.
* 두 경우 모두 결제 전 주문 정보를 입력합니다.
* 주문 정보 입력 시 기본 배송지가 있으면, 표시됩니다.
* 결제가 완료되었다는 가정 하에, 입력한 정보로 주문을 완료합니다.
###  🧷 변경사항
<!--변경 내용을 적어주세요 (커밋 번호를 적어주세요)-->
주문 정보 입력 전 필요한 정보를 가져오는 API 구현
- 4445eb05b36ce978960a61e71ace595efbf0cce9: 주문 상품 정보 가져오기 API 구현
- 80a9959952712d828e9054d0327b7a17f4971163: 기본 배송지 가져오기 API 구현
  - 배송지 입력 API도 구현했습니다.
- 76ac476498bc94ce3690eef6c1fcc0411ed2da2c: 주문 완료 후 주문 정보를 저장하는 API 구현
- 9c2d5bac0636f68699ac336341e6686bfd25f85b, de97e3638364e2c8fc9b5cbca9b59d5b8c8d7224: 테스트 작성
## 📷 스크린샷

## 👀 기타
- 테스트 커버리지 측정은 #23 과 병합 후 측정 예정

yoonnable committed:
feat: Order entity, MVC 레이아웃 추가

yoonnable committed:
fix: 불필요한 클래스 삭제

- Session 로그인 구현 시 사용한 interceptor와 설정 파일 삭제

yoonnable committed:
feat: 주문 정보 입력 전처리(주문 상품 정보 가져오기) 구현

yoonnable committed:
feat: 주문 정보 입력 전처리(기본 배송지 가져오기) 구현

- Addressbook entity, MVC 레이아웃 추가
- Access token으로 로그인한 회읜 조회 구현
- 배송지 추가 구현

yoonnable committed:
feat: 주문 기능 구현

yoonnable committed:
test: Addressbook 비즈니스로직 테스트 작성

- AddressbookRequest 테스트에 필요한 생성자 추가

yoonnable committed:
test: Order 비즈니스로직 테스트 작성

- dto 테스트에 필요한 생성자 추가

f-lab-jason commented on src/main/java/com/ecommerce/addressbook/repository/AddressbookRepository.java:
```diff
@@ -0,0 +1,12 @@
+@Repository
+public interface AddressbookRepository extends JpaRepository<Addressbook, Long> {
+
+  Addressbook findByMemberIdAndDefaultValue(Long memberId, Integer defaultValue);
```
defaultValue는 뭘 나타내는건가요..?

f-lab-jason commented on src/main/java/com/ecommerce/utils/MemberUtil.java:
```diff
@@ -0,0 +1,23 @@
+  private final JwtUtil jwtUtil;
+  private final MemberService memberService;
+
+  public Member getLoginMember(HttpServletRequest request) {
```
지금은 구조가 accessToken 발급시에 db 조회를 하고 로그인시에 해당 email로 db조회를 또 하는걸로 보이는데, 애초에 accessToken 발급시에 유저 상세정보도 같이 있는게 더 좋을 것 같은데 어떻게 생각하시나요?
지금 구조로는 요청이 갈때마다 db에서 유저를 다시 조회해야하는데 jwt를 쓰는 장점이 별로 없어보여요.

f-lab-jason commented on src/main/java/com/ecommerce/addressbook/service/AddressbookService.java:
```diff
@@ -0,0 +1,37 @@
+        .name(addressbookRequest.getName())
+        .address(addressbookRequest.getAddress())
+        .phone(addressbookRequest.getPhone())
+        .defaultValue(addressbookRequest.getDefaultValue())
```
defaultValue가 isDefault같은 boolean의 역할을 한다고 생각하는데..중복같은게 발생할 가능성이 있지 않나요? index 레벨에서 필터링 되는건지 아니면 다른 방법을 생각하시는건지 궁금해요.

f-lab-jason commented on src/main/java/com/ecommerce/order/controller/OrderController.java:
```diff
@@ -0,0 +1,72 @@
+   */
+  @GetMapping("/form")
+  public ResponseEntity<List<OrderDetailFormResponse>> orderForm(
+      @RequestBody OrderDetailFormRequest... requests) {
```
Request를 여러개 받는 개념보다는 request 안에 여러개의 주문 상품이 있는 형태인데..그러면 request 안에 list형태로 목록을 가지고 있는게 더 자연스러울 것 같아요.

f-lab-jason commented on src/main/java/com/ecommerce/order/controller/OrderController.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  /**
+   * 주문 정보 입력 전처리 (주문 상품 정보 가져오기)
```
'전처리'가 어떤 용도인가요..?

f-lab-jason commented on src/main/java/com/ecommerce/order/entity/Status.java:
```diff
@@ -0,0 +1,27 @@
+package com.ecommerce.order.entity;
+
+public enum Status {
+  PENDING,    // 주문 대기 중
+  PROCESSING, // 주문 처리 중
+  SHIPPED,    // 주문 배송 중
+  DELIVERED,  // 주문 배송 완료
+  CANCELED;   // 주문 취소
+
+  @Override
+  public String toString() {
+    switch (this) {
+      case PENDING:
+        return "Order is pending";
+      case PROCESSING:
+        return "Order is being processed";
+      case SHIPPED:
+        return "Order has been shipped";
+      case DELIVERED:
+        return "Order has been delivered";
+      case CANCELED:
+        return "Order has been canceled";
+      default:
+        return "Unknown status";
+    }
+  }
+}
```
1. OrderStatus 같이 조금 명확한 이름이면 더 좋을 것 같아요. 애플리케이션상에 수 많은 Status가 있을 수 있으니까요.
2. 아래와 같은 형태가 조금 더 자연스러워 보이는데 어떻게 생각하시나요?
```
public enum OrderStatus {
    PENDING("Order is pending"),
    ...
    CANCELED("Order has been canceled");

    private final String message;

    OrderStatus(String message) {
        this.message = message;
    }

    public String toString() {
        return this.message;
    }
}
```

yoonnable committed:
refactor: Rename Status class to OrderStatus

yoonnable committed:
refactor: OrderStatus 리팩토링

yoonnable committed:
refactor: '주문 정보 입력' 메소드 설명 주석 수정

yoonnable committed:
refactor: 주문 상품 가져오기 request 형식 리팩토링

yoonnable committed:
fix: order_detail에 order_id 자동 저장 해결

- orderDetail에 order 넣어주기
- Order의 orderDetails null 방지 처리

yoonnable committed:
refactor: 기본 배송지 설정 필드명 변경, jwt를 활용한 회원 정보 사용 리팩토링

- 기본 배송지 설정하는 필드명을 defaultValue -> isDefault 로 변경
- 서비스 사용 시 필요한 member 정보를 DB가 아닌 jwt에 담아서 사용할 수 있도록 리팩토링
- 리팩토링으로 인해 생긴 버그 해결(토큰 생성 로직)
- Order에 불필요한 연관관계 제거

yoonnable committed:
test: 리팩토링으로 인한 테스트코드 수정, 기본 배송지 리셋 테스트 코드 작성

- member 필드명 memberId로 변경
- defaultValue 필드명 isDefault로 변경
- 주문 상품 정보 요청값 타입 변환

f-lab-jason commented on src/main/java/com/ecommerce/addressbook/controller/AddressbookController.java:
```diff
@@ -23,18 +22,19 @@
   public ResponseEntity<Addressbook> getDefaultAddress(HttpServletRequest request) {
-    Member member = memberUtil.getLoginMember(request);
-    Addressbook addressbook = addressbookService.findDefaultAddressByMemberId(member.getId());
+    String accessToken = request.getHeader("Access");
+    Long memberId = jwtUtil.getMemberId(accessToken);
```
nit: 이 부분도 한 줄로 줄이고 request를 넘겨주는 형태면 더 좋을 것 같아요. 이 코드가 다른데서도 쓰일텐데, jwt에서 memberId를 가져오려면 "Access"라는 헤더명도 알아야하니까요.
`Long memberId = jwtUtil.getMemberId(request);`같은 한 줄로 처리하는게 좋을 것 같아요.

f-lab-jason commented on src/main/java/com/ecommerce/addressbook/repository/AddressbookRepository.java:
```diff
@@ -2,11 +2,17 @@
 
-  Addressbook findByMemberIdAndDefaultValue(Long memberId, Integer defaultValue);
+  @Modifying
+  @Query("update Addressbook a set a.isDefault = 0 where a.memberId = :memberId")
```
쿼리 where에 `isDefault = 1`도 있는게 좋을 것 같아요. 지금은 모든 row를 업데이트할테니까요.

f-lab-jason commented on src/main/java/com/ecommerce/addressbook/entity/Addressbook.java:
```diff
@@ -41,5 +35,5 @@ public class Addressbook extends BaseTimeEntity {
 
   @NotNull
-  private Integer defaultValue;
+  private Integer isDefault; // 0: NO, 1: YES
```
boolean이 아니라 Integer를 쓰는 이유가 있을까요?

f-lab-jason reviewed (approved):
코멘트로 남긴 간단한 내용 몇가지만 더 고려하면 좋겠어요. 고생하셨습니다 🙂

yoonnable committed:
refactor: isDefault가 1인 경우에만 업데이트하도록 쿼리 수정

- isDefault가 1인 경우에만 0으로 업데이트하도록 변경
- 불필요한 업데이트를 줄이고 쿼리 효율성 개선

yoonnable committed:
refactor: Access 토큰 처리 로직을 HeaderUtil 클래스로 통합

- HttpServletRequest에서 Access 토큰을 가져오는 getAccessToken 메서드 추가
- HttpServletResponse에 Access 토큰을 설정하는 setAccessToken 메서드 추가
- 헤더 설정을 위한 setHeader 메서드 추가
- 중복된 코드를 제거하고 코드의 재사용성을 높임

yoonnable committed:
Merge branch 'develop' into 26-feat-order

# Conflicts:
#	src/main/java/com/ecommerce/jwt/JwtFilter.java
#	src/main/java/com/ecommerce/jwt/JwtLogoutHandler.java
#	src/main/java/com/ecommerce/jwt/JwtUtil.java
#	src/main/java/com/ecommerce/jwt/SigninFilter.java
#	src/main/java/com/ecommerce/jwt/service/ReissueService.java
#	src/main/java/com/ecommerce/member/dto/MemberDetails.java

yoonnable committed:
test: 토큰 생성 및 memberId 가져오는 코드 수정

- 충돌처리

yoonnable committed:
test: 로그인 인증 성공 후 토큰 생성 테스트 코드 수정

- 토큰 생성 시 memberId, role값 추가 적용

yoonnable committed:
test: 배송지 등록 테스트 코드 커버리지 개선

- 기본 배송지 등록 시 isDefault 값에 따른 분기처리 테스트 코드 추가 작성
